### PR TITLE
IOS HLE: Replace broken path traversal prevention

### PR DESCRIPTION
**Pull request by JosJuice**
This reverts commit 141f3bfb3acee73569486258c21e4ed34fc32327.
The implementation of getting absolute paths wasn't working
on non-Windows systems, which is a huge problem for IOS HLE.

---
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/4472)

### CODE REVIEW

**Comment by JosJuice** [comment]
If someone knows how to fix this, please let me know so we don't have to merge this revert that makes Dolphin less secure.

**Comment by JosJuice** [comment]
(20:07:25) leoetlino: oh, I think I know why
(20:07:52) leoetlino: realpath() fails for files which don't exist
(20:08:35) JosJuice: Oh
(20:12:34) JosJuice: I can't find any way to get the absolute path of a file that doesn't exist :/
(20:12:49) JosJuice: There are probably ways to solve this other than getting an absolute path, but I can't think of anything robust
(20:13:53) JosJuice: The original problem is "how to check if a path points to a place outside of the NAND directory"

**Comment by hthh** [comment]
I would implement the normalisation in your own code. It might not be the best option. But the idea is roughly: replace all backslashes with forward-slashes first for Windows safety, split it into components (split on '/'), and every time you hit a ".." remove the previous component (if there is one).

The Python normpath implementation is a more thorough reference: https://github.com/python/cpython/blob/d40776b437072a3619a2b4fbfab5946efbbbc2fb/Lib/posixpath.py#L268

This all assumes the expected IOS behaviour is posix-like, which I'm not sure about.

**Comment by leoetlino** [comment]
For the reference, here's what IOS_Open returns for different paths (the canonical path being /shared2/sys/SYSCONF):

 <leoetlino> ios_open(path=/shared2/sys/SYSCONF) = 4
 <leoetlino> ios_open(path=/shared2/sys/SYSCON) = -106
 <leoetlino> ios_open(path=/shared2/sys/SYSCONF) = 4
 <leoetlino> ios_open(path=/../shared2/sys/SYSCONF) = -106
 <leoetlino> ios_open(path=/../../shared2/sys/SYSCONF) = -106
 <leoetlino> ios_open(path=/shared2/sys/../sys/SYSCONF) = -106
 <leoetlino> ios_open(path=/shared2/sys../sys/SYSCONF) = -106
 <leoetlino> ios_open(path=/shared2/sys\../sys/SYSCONF) = -106
 <leoetlino> ios_open(path=\shared2\sys\SYSCONF) = -6

So it looks like IOS doesn't have any special handling for /../, since the paths with /../ (which would be valid if it were POSIX-like) return -106.

\shared2\sys\SYSCONF returns -6 probably because it starts with \ and not with /.

**Comment by JosJuice** [comment]
Since leoetlino's testing shows that `..` appears to have no special meaning in IOS, I've decided to make a new commit that escapes names like that. It should prevent path traversal without requiring us to implement an absolute path function. The commit also improves the handling of escape sequences in file names in general.

I haven't tested it much. @JMC47, can you help us out?

**Comment by leoetlino** [comment]
Reviewed 2 of 3 files at r1, 9 of 9 files at r2, 3 of 3 files at r3.
Review status: 9 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/Common/FileUtil.cpp, line 33 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXW9vNe0moxIAbCwdQV:-KXW9vNe0moxIAbCwdQW:b-7truaf) ([raw file](https://github.com/dolphin-emu/dolphin/blob/151997336464890d7b1af24b0a6a62064a057648/Source/Core/Common/FileUtil.cpp#L33)):*
> ```cpp
> #include <errno.h>
> #include <libgen.h>
> #include <stdlib.h>
> ```

Is this include still needed?

---

*[Source/Core/Common/NandPaths.cpp, line 124 at r3](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXWBBrjrRYKGNWLENUp:-KXWBBrjrRYKGNWLENUq:b-n4z1kj) ([raw file](https://github.com/dolphin-emu/dolphin/blob/0d5c07d806883305fffdd035bdec39eabd133a5d/Source/Core/Common/NandPaths.cpp#L124)):*
> ```cpp
> std::string EscapeFileName(const std::string& filename)
> {
>   // Prevent paths from containing ./, ../, .../, ..../, and so on
> ```

Assuming this comment describes the ReplaceAll just below, I think it'd be more correct to say it prevents paths from containing only `.` (`/` doesn't seem to be checked).

Also, this is necessary because trailing `..`s can cause path traversal, is that right?

---

*[Source/Core/Common/StringUtil.h, line 109 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXW8-KlxvCP_4EF0i6z:-KXW8-KlxvCP_4EF0i7-:b-bizgxs) ([raw file](https://github.com/dolphin-emu/dolphin/blob/151997336464890d7b1af24b0a6a62064a057648/Source/Core/Common/StringUtil.h#L109)):*
> ```objc
> 
> void SplitString(const std::string& str, char delim, std::vector<std::string>& output);
> std::string UnsplitString(const std::vector<std::string>& strings, char delim);
> ```

I would call this `JoinString` (feel free to ignore)

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_FileIO.cpp, line 22 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXW99tT9-5DTVDaDIkP:-KXW99tT9-5DTVDaDIkQ:b48msni) ([raw file](https://github.com/dolphin-emu/dolphin/blob/151997336464890d7b1af24b0a6a62064a057648/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_FileIO.cpp#L22)):*
> ```cpp
> 
> // This is used by several of the FileIO and /dev/fs functions
> std::string HLE_IPC_BuildFilename(std::string wii_path)
> ```

Should this take a `const std::string&` instead?

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_FileIO.cpp, line 25 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXW8RVBhbrdjiwnR4QU:-KXW8RVBhbrdjiwnR4QV:b-p8t07p) ([raw file](https://github.com/dolphin-emu/dolphin/blob/151997336464890d7b1af24b0a6a62064a057648/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_FileIO.cpp#L25)):*
> wii_path[0] != '/'
Is this check actually needed? Or do we just fall back to FileIO every time we can't find a matching device path? If so, I think this check should be moved in WII_IPC_HLE instead to not use FileIO in the first place if the first character isn't / (to match IOS behaviour).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472)*

**Comment by JosJuice** [comment]
Review status: 5 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/Common/FileUtil.cpp, line 33 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXW9vNe0moxIAbCwdQV:-KXWD_XHOe1XDY5H5ETf:b-i2ii) ([raw file](https://github.com/dolphin-emu/dolphin/blob/151997336464890d7b1af24b0a6a62064a057648/Source/Core/Common/FileUtil.cpp#L33)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> Is this include still needed?
</details>
Probably.

---

*[Source/Core/Common/NandPaths.cpp, line 124 at r3](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXWBBrjrRYKGNWLENUp:-KXWDgTXAH8EfubyzhTZ:b-ai18zj) ([raw file](https://github.com/dolphin-emu/dolphin/blob/0d5c07d806883305fffdd035bdec39eabd133a5d/Source/Core/Common/NandPaths.cpp#L124)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> Assuming this comment describes the ReplaceAll just below, I think it'd be more correct to say it prevents paths from containing only `.` (`/` doesn't seem to be checked).
> 
> Also, this is necessary because trailing `..`s can cause path traversal, is that right?
</details>
The comment doesn't say it prevents filenames from being `../`, it says it prevents paths from containing `../`. EscapePath will filter out every `/` before calling EscapeFileName. I might as well remove the slashes from the comment, though.

`...`, `....` and so on have special meanings at least on Windows. On my system, they seem to act like `.`, but I've read somewhere that `...` can act as `..\..` and `....` can act as `..\..\..` on some systems. Even if it's not necessary for preventing path traversal, it's necessary for ending up in the directory that the emulated software expects.

---

*[Source/Core/Common/StringUtil.h, line 109 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXW8-KlxvCP_4EF0i6z:-KXWEo-1hPzQRMpXzz2M:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/151997336464890d7b1af24b0a6a62064a057648/Source/Core/Common/StringUtil.h#L109)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> I would call this `JoinString` (feel free to ignore)
</details>
Done.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_FileIO.cpp, line 22 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXW99tT9-5DTVDaDIkP:-KXWEuHwDcysvVRKkJS5:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/151997336464890d7b1af24b0a6a62064a057648/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_FileIO.cpp#L22)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> Should this take a `const std::string&` instead?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472)*

**Comment by sepalani** [comment]
Concerning the SplitString usage. I guess if the path contains an escaped backslash it will mistakenly split it which might be the wrong behavior. If that case doesn't exist then feel free to ignore this.

**Comment by JosJuice** [comment]
IOS probably doesn't support escaping like that, but it hasn't been tested.

File: Source/Core/Common/StringUtil.cpp [hunk 1]
```diff
@@ -328,6 +328,25 @@ void SplitString(const std::string& str, const char delim, std::vector<std::stri
   output.pop_back();
 }
 
+std::string JoinString(const std::vector<std::string>& strings, char delim)
```
Comment (lioncash) [inline Source/Core/Common/StringUtil.cpp]: Joining strings can be done like so, which also allows for joining via delimeters consisting of more than one char:
```cpp
std::string JoinStrings(const std::vector<std::string>& strings, const std::string& delimeter)
{
    std::stringstream res;
    std::copy(strings.begin(), strings.end(),
              std::ostream_iterator<string>(res, delimeter.c_str()));

    // Drop the trailing delimeter.
    std::string joined = res.str();
    return joined.substr(0, joined.length() - delimeter.length());
}
```

**Comment by JosJuice** [comment]
Review status: 4 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_FileIO.cpp, line 25 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXW8RVBhbrdjiwnR4QU:-KXWiK_Uj3LqjgWUduIH:b-b2im8z) ([raw file](https://github.com/dolphin-emu/dolphin/blob/151997336464890d7b1af24b0a6a62064a057648/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_FileIO.cpp#L25)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> > wii_path[0] != '/'
> Is this check actually needed? Or do we just fall back to FileIO every time we can't find a matching device path? If so, I think this check should be moved in WII_IPC_HLE instead to not use FileIO in the first place if the first character isn't / (to match IOS behaviour).
</details>
I have added checks so that this function only is called when the path starts with /.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472)*

**Comment by JosJuice** [comment]
Review status: 4 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/Common/StringUtil.cpp, line 331 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXW5zCAmDUA6ZGYZ5BF-r4-331:-KXWvauN2e8Yp0YoAFfy:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/31e7a9647b61780a1b105df3ec7bd94b92dbd73c/Source/Core/Common/StringUtil.cpp#L331)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Joining strings can be done like so, which also allows for joining via delimeters consisting of more than one char:
> 
> ```cpp
> 
> std::string JoinStrings(const std::vector<std::string>& strings, const std::string& delimeter)
> 
> {
> 
>   if (strings.empty())
> 
>     return "";
> 
> 
> 
>   std::stringstream res;
> 
>   std::copy(strings.begin(), strings.end(),
> 
>             std::ostream_iterator<std::string>(res, delimeter.c_str()));
> 
> 
> 
>   // Drop the trailing delimeter.
> 
>   std::string joined = res.str();
> 
>   return joined.substr(0, joined.length() - delimeter.length());
> 
> }
> 
> ```
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472)*

**Comment by leoetlino** [comment]
Reviewed 2 of 6 files at r4, 3 of 3 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472)*

**Comment by leoetlino** [comment]
LGTM. I also tested this on Linux, and it fixes the huge regression introduced by #4447.

**Comment by JosJuice** [comment]
Don't merge this yet. I'm writing unit tests, and they seem to have uncovered errors in the escaping.

**Comment by JosJuice** [comment]
Errors fixed. Everything should be working correctly now.

**Comment by lioncash** [comment]
Review status: 11 of 14 files reviewed at latest revision, 13 unresolved discussions.

---

*[Source/Core/Common/FileUtil.cpp, line 33 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXV6iWUK1V5OY5oRHp:-KXXV6iWUK1V5OY5oRHq:bp914mm) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Common/FileUtil.cpp#L33)):*
> ```cpp
> #include <errno.h>
> #include <libgen.h>
> #include <stdlib.h>
> ```

Why was this changed from `<cstdlib>` ?

---

*[Source/Core/Common/StringUtil.cpp, line 331 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXW5zCAmDUA6ZGYZ5BF-r4-331:-KXXVMu3N1-CJlYUJtlS:b344bct) ([raw file](https://github.com/dolphin-emu/dolphin/blob/31e7a9647b61780a1b105df3ec7bd94b92dbd73c/Source/Core/Common/StringUtil.cpp#L331)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Done.
</details>
I edited the example on Github but forgot they don't reflect back on Reviewable, but you might want to add

```cpp
if (strings.empty())
  return "";
```

before the stringstream declaration so it doesn't construct a bunch of other stuff and can just early exit.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE.cpp, line 444 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXWIJYWIHUPfwHphJg:-KXXWIJYWIHUPfwHphJh:b20482u) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE.cpp#L444)):*
> ```cpp
>         }
>       }
>       else if (device_name.find("/") == 0)
> ```

since string's find returns the index of the character, this can just be:

```
else if (device_name.front() == '/')
{
```

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_FileIO.cpp, line 26 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXWyIyaz6Qo1Zy4EMc:-KXXWyIzMMF9h99BniRy:b-g6i4yj) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_FileIO.cpp#L26)):*
> ```cpp
> {
>   std::string nand_path = File::GetUserPath(D_SESSION_WIIROOT_IDX);
>   if (wii_path.compare(0, 1, "/") == 0)
> ```

This can just be:

```cpp
if (wii_path.front() == '/')
```

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp, line 99 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXX61_FD83kqgzYsVi:-KXXX61a8fq-Ca4W110f:b-2b5ou0) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp#L99)):*
> ```cpp
>         Memory::GetString(CommandBuffer.InBuffer[0].m_Address, CommandBuffer.InBuffer[0].m_Size);
> 
>     if (relative_path.compare(0, 1, "/") != 0)
> ```

This can just be:

```cpp
if (relative_path.front() != '/')
```

I've listed similar ones below, but you may just want to make a local function names `IsValidRelativePath()` (use a better name if applicable) that just does this check like:

```cpp
bool IsValidRelativePath(const std::string& path)
{
    return path.front() == '/';
}
```

So it keeps the sanitizing check in one place.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp, line 190 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXXKUYpl2HBgtUsSiw:-KXXXKUZlcL_KO1WeU9d:b8m70kt) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp#L190)):*
> ```cpp
>         Memory::GetString(CommandBuffer.InBuffer[0].m_Address, CommandBuffer.InBuffer[0].m_Size);
> 
>     if (relativepath.compare(0, 1, "/") != 0)
> ```

This can just be:

```cpp
if (relativepath.front() != '/')
```

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp, line 312 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXXQURZYOhZ_DU6ZQp:-KXXXQURZYOhZ_DU6ZQq:bylaki1) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp#L312)):*
> ```cpp
>     Addr += 2;
>     const std::string wii_path = Memory::GetString(Addr, 64);
>     if (wii_path.compare(0, 1, "/") != 0)
> ```

This can just be:

```cpp
if (wii_path.front() != '/')
```

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp, line 343 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXXXov6qvhiMpFaN_D:-KXXXXov6qvhiMpFaN_E:bylaki1) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp#L343)):*
> ```cpp
>     Addr += 2;
>     const std::string wii_path = Memory::GetString(_BufferIn, 64);
>     if (wii_path.compare(0, 1, "/") != 0)
> ```

This can just be:

```cpp
if (wii_path.front() != '/')
```

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp, line 381 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXXc_erzBQx8l1L_mg:-KXXXc_erzBQx8l1L_mh:bylaki1) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp#L381)):*
> ```cpp
>                            // (0x3038) for MH3 etc
>     const std::string wii_path = Memory::GetString(_BufferIn, 64);
>     if (wii_path.compare(0, 1, "/") != 0)
> ```

This can just be:

```cpp
if (wii_path.front() != '/')
```

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp, line 440 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXXdTIq04VBOasOcNG:-KXXXdTIq04VBOasOcNH:bylaki1) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp#L440)):*
> ```cpp
> 
>     const std::string wii_path = Memory::GetString(_BufferIn + Offset, 64);
>     if (wii_path.compare(0, 1, "/") != 0)
> ```

This can just be:

```cpp
if (wii_path.front() != '/')
```

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp, line 470 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXXfAdDH3OQbHT-Sbt:-KXXXfAdDH3OQbHT-Sbu:bylaki1) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp#L470)):*
> ```cpp
> 
>     const std::string wii_path = Memory::GetString(_BufferIn + Offset, 64);
>     if (wii_path.compare(0, 1, "/") != 0)
> ```

This can just be:

```cpp
if (wii_path.front() != '/')
```

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp, line 479 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXXixu-2bB8c0S5EiW:-KXXXixu-2bB8c0S5EiX:br2wc7q) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp#L479)):*
> ```cpp
> 
>     const std::string wii_path_rename = Memory::GetString(_BufferIn + Offset, 64);
>     if (wii_path_rename.compare(0, 1, "/") != 0)
> ```

This can just be:

```cpp
if (wii_path_rename.front() != '/')
```

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp, line 522 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXXhVCxy1XhnV4SVua:-KXXXhVCxy1XhnV4SVub:bylaki1) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp#L522)):*
> ```cpp
>     Addr += 2;
>     const std::string wii_path = Memory::GetString(Addr, 64);
>     if (wii_path.compare(0, 1, "/") != 0)
> ```

This can just be:

```cpp
if (wii_path.front() != '/')
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472)*

**Comment by JosJuice** [comment]
Review status: 11 of 14 files reviewed at latest revision, 13 unresolved discussions.

---

*[Source/Core/Common/FileUtil.cpp, line 33 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXV6iWUK1V5OY5oRHp:-KXXZzTbr1KiihJ7H3Lm:b-9xluq4) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Common/FileUtil.cpp#L33)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Why was this changed from `<cstdlib>` ?
</details>
It was `stdlib.h` before the broken PR, and the first commit of this PR is a pure revert of the broken PR. Should I edit the revert commit to turn this into `cstdlib`?

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE.cpp, line 444 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXWIJYWIHUPfwHphJg:-KXX_SGdn9nTDoJIwWAT:b6ois63) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE.cpp#L444)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> since string's find returns the index of the character, this can just be:
> 
> ```
> else if (device_name.front() == '/')
> {
> ```
</details>
If the string is empty, `front` won't work, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472)*

**Comment by lioncash** [comment]
Review status: 11 of 14 files reviewed at latest revision, 13 unresolved discussions.

---

*[Source/Core/Common/FileUtil.cpp, line 33 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXV6iWUK1V5OY5oRHp:-KXXayIPMA4rFDleS81M:btmnpjz) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Common/FileUtil.cpp#L33)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> It was `stdlib.h` before the broken PR, and the first commit of this PR is a pure revert of the broken PR. Should I edit the revert commit to turn this into `cstdlib`?
</details>
Ah, ok, nevermind then.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE.cpp, line 444 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXWIJYWIHUPfwHphJg:-KXXadQV4SuyhIWytA29:bfddbwd) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE.cpp#L444)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> If the string is empty, `front` won't work, right?
</details>
Ah, that's true. Ignore the rest of the comments changing it to `front()` then. However, with find, you can just a character literal instead of a string literal.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472)*

**Comment by JosJuice** [comment]
Review status: 8 of 14 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/Common/StringUtil.cpp, line 331 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXW5zCAmDUA6ZGYZ5BF-r4-331:-KXXdbBjUrLJ7oNYkl-2:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/31e7a9647b61780a1b105df3ec7bd94b92dbd73c/Source/Core/Common/StringUtil.cpp#L331)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> I edited the example on Github but forgot they don't reflect back on Reviewable, but you might want to add
> 
> ```cpp
> if (strings.empty())
>   return "";
> ```
> 
> before the stringstream declaration so it doesn't construct a bunch of other stuff and can just early exit.
</details>
Done.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE.cpp, line 444 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXWIJYWIHUPfwHphJg:-KXXeQmCdOErNPF4UPbN:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE.cpp#L444)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Ah, that's true. Ignore the rest of the comments changing it to `front()` then. However, with find, you can just a character literal instead of a string literal.
</details>
Done.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp, line 99 at r8](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXXX61_FD83kqgzYsVi:-KXXezasTcsEY1q54aVc:b-kh1du3) ([raw file](https://github.com/dolphin-emu/dolphin/blob/531b5c3d2257cffd57cbbe7300974f680bef5164/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp#L99)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This can just be:
> 
> ```cpp
> if (relative_path.front() != '/')
> ```
> 
> I've listed similar ones below, but you may just want to make a local function names `IsValidRelativePath()` (use a better name if applicable) that just does this check like:
> 
> ```cpp
> bool IsValidRelativePath(const std::string& path)
> {
>     return path.front() == '/';
> }
> ```
> 
> So it keeps the sanitizing check in one place.
</details>
I've made a local function for it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472)*

**Comment by leoetlino** [comment]
Reviewed 5 of 5 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472)*

**Comment by leoetlino** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp, line 25 at r10](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXb_VTZoGsMentse3ob:-KXb_VTZoGsMentse3oc:b-7v8r65) ([raw file](https://github.com/dolphin-emu/dolphin/blob/133c2a7b5e2594bd19014cae611ab3cce479032f/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp#L25)):*
> ```cpp
> #include "Core/IPC_HLE/WII_IPC_HLE_Device_fs.h"
> 
> bool IsValidWiiPath(const std::string& path)
> ```

This should probably be `static`. (GCC warns of a missing declaration.)

---

*[Source/UnitTests/Common/NandPathsTest.cpp, line 10 at r10](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXb_gjqkyDjpR1TH4dV:-KXb_gjrFCcKVxubIk-G:b-7v8r65) ([raw file](https://github.com/dolphin-emu/dolphin/blob/133c2a7b5e2594bd19014cae611ab3cce479032f/Source/UnitTests/Common/NandPathsTest.cpp#L10)):*
> ```cpp
> #include "Common/NandPaths.h"
> 
> void TestEscapeAndUnescape(const std::string& unescaped, const std::string& escaped)
> ```

This should probably be `static`. (GCC warns of a missing declaration.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472)*

**Comment by JosJuice** [comment]
Review status: 13 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp, line 25 at r10](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXb_VTZoGsMentse3ob:-KXbdz0fLblDW5mK8JUx:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/133c2a7b5e2594bd19014cae611ab3cce479032f/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_fs.cpp#L25)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> This should probably be `static`. (GCC warns of a missing declaration.)
</details>
Done.

---

*[Source/UnitTests/Common/NandPathsTest.cpp, line 10 at r10](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472#-KXb_gjqkyDjpR1TH4dV:-KXbe-AsZGuAiueSeQMh:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/133c2a7b5e2594bd19014cae611ab3cce479032f/Source/UnitTests/Common/NandPathsTest.cpp#L10)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> This should probably be `static`. (GCC warns of a missing declaration.)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472)*

**Comment by leoetlino** [comment]
Reviewed 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4472)*

**Comment by JosJuice** [comment]
Since the issue that currently is in master is severe, I will merge this PR myself after 24 hours from now unless anyone objects to it (or unless anyone wants to merge it on their own before that).

**Pull request merged**